File: src/nodetool/models/condition_builder.py
from typing import Any, List, Union
from enum import Enum


class Operator(Enum):
    EQ = "="
    NE = "!="
    GT = ">"
    LT = "<"
    GTE = ">="
    LTE = "<="
    IN = "IN"
    LIKE = "LIKE"


class LogicalOperator(Enum):
    AND = "AND"
    OR = "OR"


class Variable:
    def __init__(self, name: str):
        self.name = name


class Condition:
    def __init__(self, field: str, operator: Operator, value: Any):
        self.field = field
        self.operator = operator
        self.value = value


class ConditionGroup:
    def __init__(
        self,
        conditions: List[Union["ConditionGroup", Condition]],
        operator: LogicalOperator,
    ):
        self.conditions = conditions
        self.operator = operator


class Field:
    def __init__(self, name: str):
        self.name = name

    def _create_condition(
        self, operator: Operator, value: Union[Any, Variable]
    ) -> "ConditionBuilder":
        return ConditionBuilder(Condition(self.name, operator, value))

    def equals(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.EQ, value)

    def not_equals(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.NE, value)

    def greater_than(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.GT, value)

    def less_than(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.LT, value)

    def greater_than_or_equal(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.GTE, value)

    def less_than_or_equal(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.LTE, value)

    def in_list(self, values: Union[List[Any], Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.IN, values)

    def like(self, pattern: Union[str, Variable]) -> "ConditionBuilder":
        return self._create_condition(Operator.LIKE, pattern)

    def __eq__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.equals(value)

    def __ne__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.not_equals(value)

    def __gt__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.greater_than(value)

    def __lt__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.less_than(value)

    def __ge__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.greater_than_or_equal(value)

    def __le__(self, value: Union[Any, Variable]) -> "ConditionBuilder":
        return self.less_than_or_equal(value)

    def __contains__(self, value: Union[List[Any], Variable]) -> "ConditionBuilder":
        return self.in_list(value)


class ConditionBuilder:
    def __init__(self, condition: Union[Condition, ConditionGroup]):
        self.root = ConditionGroup([condition], LogicalOperator.AND)

    def _add_condition(
        self, other: "ConditionBuilder", operator: LogicalOperator
    ) -> "ConditionBuilder":
        if (
            isinstance(self.root.conditions[0], Condition)
            and len(self.root.conditions) == 1
        ):
            self.root.conditions = [self.root.conditions[0], other.root.conditions[0]]
            self.root.operator = operator
        else:
            new_root = ConditionGroup([self.root, other.root], operator)
            self.root = new_root
        return self

    def and_(self, other: "ConditionBuilder") -> "ConditionBuilder":
        return self._add_condition(other, LogicalOperator.AND)

    def or_(self, other: "ConditionBuilder") -> "ConditionBuilder":
        return self._add_condition(other, LogicalOperator.OR)

    def build(self) -> ConditionGroup:
        return self.root

    def _get_variables(
        self, values: dict[str, Any], condition: Union[Condition, ConditionGroup]
    ):
        if isinstance(condition, Condition):
            if isinstance(condition.value, Variable):
                values[condition.value.name] = None
        else:
            for sub_condition in condition.conditions:
                self._get_variables(values, sub_condition)

    def get_variables(self) -> dict[str, Any]:
        values = {}
        self._get_variables(values, self.root)
        return values


# Example usage:
# condition = Field("age").greater_than(Variable("min_age")).and_(Field("name").like(Variable("name_pattern")))
# built_condition = condition.build()

File: packages/models/src/condition-builder.ts
/**
 * Condition Builder – fluent API for constructing query filter trees.
 *
 * Port of Python's `nodetool.models.condition_builder`.
 *
 * Example:
 * ```ts
 * const cond = field("age").greaterThan(18)
 *   .and(field("status").equals("active"))
 *   .or(field("priority").inList(["high", "critical"]));
 * ```
 */

// ── Enums ────────────────────────────────────────────────────────────

export enum Operator {
  EQ = "=",
  NE = "!=",
  GT = ">",
  LT = "<",
  GTE = ">=",
  LTE = "<=",
  IN = "IN",
  LIKE = "LIKE",
  CONTAINS = "CONTAINS",
  IS_NULL = "IS NULL",
  IS_NOT_NULL = "IS NOT NULL",
}

export enum LogicalOperator {
  AND = "AND",
  OR = "OR",
}

// ── Value Types ──────────────────────────────────────────────────────

/** A named placeholder that can be resolved at evaluation time. */
export class Variable {
  constructor(public readonly name: string) {}
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ConditionValue = any;

// ── Condition Nodes ──────────────────────────────────────────────────

/** A single comparison (e.g. `age > 18`). */
export class Condition {
  constructor(
    public readonly field: string,
    public readonly operator: Operator,
    public readonly value: ConditionValue,
  ) {}
}

/** A group of conditions joined by a logical operator. */
export class ConditionGroup {
  public readonly conditions: Array<Condition | ConditionGroup>;
  public readonly operator: LogicalOperator;

  constructor(
    conditions: Array<Condition | ConditionGroup | ConditionBuilder>,
    operator: LogicalOperator,
  ) {
    this.operator = operator;
    this.conditions = conditions.map((c) =>
      c instanceof ConditionBuilder ? c.build() : c,
    );
  }
}

// ── Field (fluent entry-point) ───────────────────────────────────────

export class Field {
  constructor(public readonly name: string) {}

  private _cond(op: Operator, value: ConditionValue): ConditionBuilder {
    return new ConditionBuilder(new Condition(this.name, op, value));
  }

  equals(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.EQ, value);
  }
  notEquals(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.NE, value);
  }
  greaterThan(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.GT, value);
  }
  lessThan(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.LT, value);
  }
  greaterThanOrEqual(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.GTE, value);
  }
  lessThanOrEqual(value: ConditionValue): ConditionBuilder {
    return this._cond(Operator.LTE, value);
  }
  inList(values: ConditionValue[] | Variable): ConditionBuilder {
    return this._cond(Operator.IN, values);
  }
  like(pattern: string | Variable): ConditionBuilder {
    return this._cond(Operator.LIKE, pattern);
  }
  isNull(): ConditionBuilder {
    return this._cond(Operator.IS_NULL, null);
  }
  isNotNull(): ConditionBuilder {
    return this._cond(Operator.IS_NOT_NULL, null);
  }
}

/** Convenience factory: `field("age").greaterThan(18)` */
export function field(name: string): Field {
  return new Field(name);
}

// ── Condition Builder (chaining) ─────────────────────────────────────

export class ConditionBuilder {
  public root: ConditionGroup;

  constructor(condition: Condition | ConditionGroup) {
    if (condition instanceof Condition) {
      this.root = new ConditionGroup([condition], LogicalOperator.AND);
    } else {
      this.root = condition;
    }
  }

  private _add(
    other: ConditionBuilder,
    op: LogicalOperator,
  ): ConditionBuilder {
    const selfIsSingle =
      this.root.conditions.length === 1 &&
      this.root.conditions[0] instanceof Condition &&
      this.root.operator === LogicalOperator.AND;

    const otherIsSingle =
      other.root.conditions.length === 1 &&
      other.root.operator === LogicalOperator.AND;

    if (selfIsSingle) {
      const otherCond = otherIsSingle
        ? other.root.conditions[0]
        : other.root;
      this.root = new ConditionGroup(
        [this.root.conditions[0], otherCond],
        op,
      );
    } else if (otherIsSingle) {
      this.root = new ConditionGroup(
        [this.root, other.root.conditions[0]],
        op,
      );
    } else {
      this.root = new ConditionGroup([this.root, other.root], op);
    }
    return this;
  }

  and(other: ConditionBuilder): ConditionBuilder {
    return this._add(other, LogicalOperator.AND);
  }

  or(other: ConditionBuilder): ConditionBuilder {
    return this._add(other, LogicalOperator.OR);
  }

  build(): ConditionGroup {
    return this.root;
  }

  /** Extract all Variable names used in the tree. */
  getVariables(): Record<string, null> {
    const vars: Record<string, null> = {};
    this._collectVars(this.root, vars);
    return vars;
  }

  private _collectVars(
    node: Condition | ConditionGroup,
    vars: Record<string, null>,
  ): void {
    if (node instanceof Condition) {
      if (node.value instanceof Variable) {
        vars[node.value.name] = null;
      } else if (Array.isArray(node.value)) {
        for (const v of node.value) {
          if (v instanceof Variable) vars[v.name] = null;
        }
      }
    } else if (node instanceof ConditionGroup) {
      for (const c of node.conditions) {
        this._collectVars(c, vars);
      }
    }
  }
}
